californium-core/src/main/java/org/eclipse/californium/core/coap/Request.java

```diff
@@ -588,21 +581,17 @@ public Response waitForResponse() throws InterruptedException {
 	public Response waitForResponse(long timeout) throws InterruptedException {
 		long before = System.currentTimeMillis();
 		long expired = timeout > 0 ? (before + timeout) : 0;
-		// Lazy initialization of a lock
 		synchronized (this) {
-			if (lock == null) {
-				lock = new Object();
-			}
-		}
-		// wait for response
-		synchronized (lock) {
 			while (this.response == null && !isCanceled() && !isTimedOut() && !isRejected()) {
-				lock.wait(timeout);
+				wait(timeout);
 				long now = System.currentTimeMillis();
 				// timeout expired?
-				if (timeout > 0 && expired <= now) {
-					// break loop since response is still null
-					break;
+				if (timeout > 0) {
+					timeout = expired - now;
```

Comment:
what's wrong with the previous condition? IMHO you shouldn't assign a value to the `timeout` parameter ...